### PR TITLE
KAFKA-9925: decorate pseudo-topics with app id

### PR DESCRIPTION
**Pull request by vvcephei**
* ensure that pseudo-topics get correctly prefixed with the app id at run time
* update test to expect the app id prefix

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 1]
```diff
@@ -972,13 +974,26 @@ boolean sendingOldValueEnabled() {
...
+        final NamedInternal renamed = new NamedInternal(joinName);
+
+        final String subscriptionTopicName = renamed.suffixWithOrElseGet(
+            "-subscription-registration",
+            builder,
+            SUBSCRIPTION_REGISTRATION
+        ) + TOPIC_SUFFIX;
 
+        // the decoration can't be performed until we have the configuration available when the app runs,
+        // so we pass Suppliers into the components, which they can call at run time
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:985]: Hopefully, this explains what's going on here.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java [hunk 2]
```diff
@@ -76,6 +78,10 @@ public void setIfUnset(final Serializer<K> defaultSerializer) {
                 throw new UnsupportedVersionException("SubscriptionWrapper version is larger than maximum supported 0x7F");
             }
 
+            if (primaryKeySerializationPseudoTopic == null) {
+                primaryKeySerializationPseudoTopic = primaryKeySerializationPseudoTopicSupplier.get();
+            }
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java:83]: This (and below) is a bit awkward.

Our requirement is not to call the supplier until after the app starts, but we can call it any time after the app starts.

The natural place would be in `configure`, but unfortunately, that method is basically useless for our internal serdes. The reason is that we previously decided that `configure` should be called externally to the DSL, but our internal serdes are constructed _internal_ to the DSL. Plus, `configure` must be called at run time (when the config is available), but by run time, we can no longer tell whether our serde is "internal" or not. So, there's no good place where we can call `configure` for our internal serdes.

I'm side-stepping the problem here by just invoking the supplier when we first need to use it, which is also at run time.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java [hunk 3]
```diff
@@ -1224,6 +1224,10 @@ private static Pattern buildPattern(final Collection<String> sourceTopics,
         return decoratedTopics;
     }
 
+    public String decoratePseudoTopic(final String topic) {
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:1227]: I'm adding a new public method for our specific use case here, to document that we should _only_ need to invoke this method publicly for "pseudo" topics.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java [hunk 4]
```diff
@@ -218,19 +220,19 @@ public void shouldUseExpectedTopicsWithSerde() {
...
         // topics our serdes serialize data for
-        assertThat(serdeScope.registeredTopics(), CoreMatchers.is(mkSet(
+        assertThat(serdeScope.registeredTopics(), is(mkSet(
             // expected pseudo-topics
-            "KTABLE-FK-JOIN-SUBSCRIPTION-REGISTRATION-0000000006-topic-fk--key",
-            "KTABLE-FK-JOIN-SUBSCRIPTION-REGISTRATION-0000000006-topic-pk--key",
-            "KTABLE-FK-JOIN-SUBSCRIPTION-REGISTRATION-0000000006-topic-vh--value",
+            applicationId + "-KTABLE-FK-JOIN-SUBSCRIPTION-REGISTRATION-0000000006-topic-fk--key",
+            applicationId + "-KTABLE-FK-JOIN-SUBSCRIPTION-REGISTRATION-0000000006-topic-pk--key",
+            applicationId + "-KTABLE-FK-JOIN-SUBSCRIPTION-REGISTRATION-0000000006-topic-vh--value",
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableKTableForeignKeyJoinScenarioTest.java:225]: This verifies the fix: the pseudo topics should also be prefixed. I should have noticed before that they weren't.

**Comment by guozhangwang** [comment]
cc @abbccdda @mjsax to take a look?

**Review by arkins — APPROVED** [review]
Many thanks for the quick turn around for the fix, with application id appended at the front, the issue should be resolved.

**Comment by vvcephei** [comment]
Thanks, @arkins ! Shame is a powerful motivator :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 1 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:985]: I understood

**Review by abbccdda — APPROVED** [review]
LGTM

**Comment by vvcephei** [comment]
Thanks, all! I'll go ahead and merge this.

**Comment by vvcephei** [comment]
cherry-picked to 2.5

**Comment by vvcephei** [comment]
cherry-picked to 2.4

**Pull request merged**